File: packages/ios/src/RustO.swift
import Foundation

public struct Point2D {
    public let x: Float
    public let y: Float
    
    public init(x: Float, y: Float) {
        self.x = x
        self.y = y
    }
}

public struct Frame {
    public let width: Float
    public let height: Float
    public let top: Float
    public let left: Float
    
    public init(width: Float, height: Float, top: Float, left: Float) {
        self.width = width
        self.height = height
        self.top = top
        self.left = left
    }
}

public struct TextResult {
    public let text: String
    public let score: Float
    public let boxPoints: [Point2D]
    public let frame: Frame
    
    public init(text: String, score: Float, boxPoints: [Point2D], frame: Frame? = nil) {
        self.text = text
        self.score = score
        self.boxPoints = boxPoints
        if let frame = frame {
            self.frame = frame
        } else {
            let minX = boxPoints.map { $0.x }.min() ?? 0.0
            let maxX = boxPoints.map { $0.x }.max() ?? 0.0
            let minY = boxPoints.map { $0.y }.min() ?? 0.0
            let maxY = boxPoints.map { $0.y }.max() ?? 0.0
            self.frame = Frame(width: maxX - minX, height: maxY - minY, top: minY, left: minX)
        }
    }
}

public struct DetectionConfig: Codable {
    public var enabled: Bool? = nil
    public var modelPath: String? = nil
    public var thresh: Float? = nil
    public var boxThresh: Float? = nil
    public var unclipRatio: Float? = nil
    public var limitSideLen: Int? = nil
    public var limitType: String? = nil
    public var useDilation: Bool? = nil

    public init(enabled: Bool? = nil, modelPath: String? = nil, thresh: Float? = nil, boxThresh: Float? = nil, unclipRatio: Float? = nil, limitSideLen: Int? = nil, limitType: String? = nil, useDilation: Bool? = nil) {
        self.enabled = enabled
        self.modelPath = modelPath
        self.thresh = thresh
        self.boxThresh = boxThresh
        self.unclipRatio = unclipRatio
        self.limitSideLen = limitSideLen
        self.limitType = limitType
        self.useDilation = useDilation
    }
}

public struct RecognitionConfig: Codable {
    public var enabled: Bool? = nil
    public var modelPath: String? = nil
    public var dictPath: String? = nil
    public var scoreThresh: Float? = nil
    public var returnWordBox: Bool? = nil
    public var returnSingleCharBox: Bool? = nil

    public init(enabled: Bool? = nil, modelPath: String? = nil, dictPath: String? = nil, scoreThresh: Float? = nil, returnWordBox: Bool? = nil, returnSingleCharBox: Bool? = nil) {
        self.enabled = enabled
        self.modelPath = modelPath
        self.dictPath = dictPath
        self.scoreThresh = scoreThresh
        self.returnWordBox = returnWordBox
        self.returnSingleCharBox = returnSingleCharBox
    }
}

/// NOTE: Available ONLY on PP-OCRv4 and PP-OCRv5
public struct ClassificationConfig: Codable {
    public var enabled: Bool? = nil
    public var modelPath: String? = nil
    public var thresh: Float? = nil

    public init(enabled: Bool? = nil, modelPath: String? = nil, thresh: Float? = nil) {
        self.enabled = enabled
        self.modelPath = modelPath
        self.thresh = thresh
    }
}

public struct OrientationConfig: Codable {
    public var enabled: Bool? = nil
    public var modelPath: String? = nil
    public var thresh: Float? = nil

    public init(enabled: Bool? = nil, modelPath: String? = nil, thresh: Float? = nil) {
        self.enabled = enabled
        self.modelPath = modelPath
        self.thresh = thresh
    }
}

public struct UnwarpConfig: Codable {
    public var enabled: Bool? = nil
    public var modelPath: String? = nil

    public init(enabled: Bool? = nil, modelPath: String? = nil) {
        self.enabled = enabled
        self.modelPath = modelPath
    }
}

public struct LayoutConfig: Codable {
    public var yThresholdMultiplier: Float? = nil
    public var xThresholdMultiplier: Float? = nil

    public init(yThresholdMultiplier: Float? = nil, xThresholdMultiplier: Float? = nil) {
        self.yThresholdMultiplier = yThresholdMultiplier
        self.xThresholdMultiplier = xThresholdMultiplier
    }
}

public struct InitializeConfig: Codable {
    public var template: String? = nil
    public var detection: DetectionConfig? = nil
    public var recognition: RecognitionConfig? = nil
    public var classification: ClassificationConfig? = nil
    public var orientation: OrientationConfig? = nil
    public var unwarp: UnwarpConfig? = nil
    public var layout: LayoutConfig? = nil

    public init(
        template: String? = nil,
        detection: DetectionConfig? = nil,
        recognition: RecognitionConfig? = nil,
        classification: ClassificationConfig? = nil,
        orientation: OrientationConfig? = nil,
        unwarp: UnwarpConfig? = nil,
        layout: LayoutConfig? = nil
    ) {
        self.template = template
        self.detection = detection
        self.recognition = recognition
        self.classification = classification
        self.orientation = orientation
        self.unwarp = unwarp
        self.layout = layout
    }

    public static func ppv6(detection: DetectionConfig? = nil, recognition: RecognitionConfig? = nil) -> InitializeConfig {
        InitializeConfig(template: "ppv6", detection: detection, recognition: recognition)
    }

    public static func ppv5(detection: DetectionConfig? = nil, recognition: RecognitionConfig? = nil) -> InitializeConfig {
        InitializeConfig(template: "ppv5", detection: detection, recognition: recognition)
    }

    public static func ppv4(detection: DetectionConfig? = nil, recognition: RecognitionConfig? = nil) -> InitializeConfig {
        InitializeConfig(template: "ppv4", detection: detection, recognition: recognition)
    }

    public static func ppv3(detection: DetectionConfig? = nil, recognition: RecognitionConfig? = nil) -> InitializeConfig {
        InitializeConfig(template: "ppv3", detection: detection, recognition: recognition)
    }
}

public enum RustOError: Error {
    case initializationFailed
    case recognitionFailed(Int32)
    case invalidHandle
    case invalidPath
    case serializationFailed
}

public enum OutputGranularity: String, Codable {
    case lines, words, spatial
}

public struct PostprocessRunOptions: Codable {
    public var threshold: Float?
    public var boxThreshold: Float?
    public var maxCandidates: Int?
    public var unclipRatio: Float?
    public var useDilation: Bool?
    public init(threshold: Float? = nil, boxThreshold: Float? = nil, maxCandidates: Int? = nil, unclipRatio: Float? = nil, useDilation: Bool? = nil) { self.threshold = threshold; self.boxThreshold = boxThreshold; self.maxCandidates = maxCandidates; self.unclipRatio = unclipRatio; self.useDilation = useDilation }
}

public struct DetectionRunOptions: Codable {
    public var limitSideLen: Int?
    public var limitType: String?
    public var mean: [Float]?
    public var std: [Float]?
    public var postprocess: PostprocessRunOptions?
    public init(limitSideLen: Int? = nil, limitType: String? = nil, mean: [Float]? = nil, std: [Float]? = nil, postprocess: PostprocessRunOptions? = nil) { self.limitSideLen = limitSideLen; self.limitType = limitType; self.mean = mean; self.std = std; self.postprocess = postprocess }
}

public struct PreprocessingRunOptions: Codable {
    public var minHeight: Float?
    public var maxSideLen: Float?
    public var minSideLen: Float?
    public var widthHeightRatio: Float?
    public var detection: DetectionRunOptions?
    public init(minHeight: Float? = nil, maxSideLen: Float? = nil, minSideLen: Float? = nil, widthHeightRatio: Float? = nil, detection: DetectionRunOptions? = nil) { self.minHeight = minHeight; self.maxSideLen = maxSideLen; self.minSideLen = minSideLen; self.widthHeightRatio = widthHeightRatio; self.detection = detection }
}

public struct OcrRunOptions: Codable {
    public var output: OutputGranularity
    public var lineYThreshold: Float?
    public var wordXThreshold: Float?
    public var textScore: Float?
    public var classification: Bool?
    public var orientation: Bool?
    public var preprocessing: PreprocessingRunOptions?

    public init(output: OutputGranularity = .lines, lineYThreshold: Float? = nil, wordXThreshold: Float? = nil, textScore: Float? = nil, classification: Bool? = nil, orientation: Bool? = nil, preprocessing: PreprocessingRunOptions? = nil) {
        self.output = output; self.lineYThreshold = lineYThreshold; self.wordXThreshold = wordXThreshold; self.textScore = textScore; self.classification = classification; self.orientation = orientation; self.preprocessing = preprocessing
    }
}

public enum ImageSource {
    case uri(String)
    case bytes(Data)
}

public enum DetectTextResult {
    case structured([TextResult])
    case spatial(String)
}

public class RustO {
    private var handle: OpaquePointer?

    private init(handle: OpaquePointer) {
        self.handle = handle
    }

    public static var version: String {
        guard let versionPtr = rocr_version() else { return "unknown" }
        return String(cString: versionPtr)
    }

    public static func initialize(config: InitializeConfig = InitializeConfig()) throws -> RustO {
        var resolvedConfig = config
        var det = config.detection ?? DetectionConfig()
        let detName = det.modelPath ?? "det.mnn"
        det.modelPath = resolveModelPath(detName) ?? detName
        resolvedConfig.detection = det

        var rec = config.recognition ?? RecognitionConfig()
        let recName = rec.modelPath ?? "rec.mnn"
        let dictName = rec.dictPath ?? "dict.txt"
        rec.modelPath = resolveModelPath(recName) ?? recName
        rec.dictPath = resolveModelPath(dictName) ?? dictName
        resolvedConfig.recognition = rec

        if var cls = config.classification, let path = cls.modelPath {
            cls.modelPath = resolveModelPath(path) ?? path
            resolvedConfig.classification = cls
        }
        if var orient = config.orientation, let path = orient.modelPath {
            orient.modelPath = resolveModelPath(path) ?? path
            resolvedConfig.orientation = orient
        }
        if var unwarp = config.unwarp, let path = unwarp.modelPath {
            unwarp.modelPath = resolveModelPath(path) ?? path
            resolvedConfig.unwarp = unwarp
        }

        guard let data = try? JSONEncoder().encode(resolvedConfig),
              let json = String(data: data, encoding: .utf8),
              let handle = rocr_initialize(json) else {
            throw RustOError.initializationFailed
        }
        return RustO(handle: handle)
    }

    public func detectText(_ source: ImageSource, options: OcrRunOptions = OcrRunOptions()) throws -> DetectTextResult {
        guard let handle else { throw RustOError.invalidHandle }
        try validate(options)
        guard let data = try? JSONEncoder().encode(options),
              let json = String(data: data, encoding: .utf8) else {
            throw RustOError.serializationFailed
        }
        switch (source, options.output) {
        case let (.uri(value), .lines), let (.uri(value), .words):
            return .structured(try structuredFile(handle, path: resolveFilePath(value), options: json))
        case let (.bytes(data), .lines), let (.bytes(data), .words):
            return .structured(try structuredData(handle, data: data, options: json))
        case let (.uri(value), .spatial):
            guard let text = rocr_detect_text_file_spatial(handle, resolveFilePath(value), json) else {
                throw RustOError.recognitionFailed(-1)
            }
            defer { rocr_free_string(text) }
            return .spatial(String(cString: text))
        case let (.bytes(data), .spatial):
            guard !data.isEmpty else { throw RustOError.invalidPath }
            let text = data.withUnsafeBytes { bytes in
                rocr_detect_text_data_spatial(handle, bytes.baseAddress, data.count, json)
            }
            guard let text else { throw RustOError.recognitionFailed(-1) }
            defer { rocr_free_string(text) }
            return .spatial(String(cString: text))
        }
    }

    private func structuredFile(_ handle: OpaquePointer, path: String, options: String) throws -> [TextResult] {
        var results: UnsafeMutablePointer<CTextResult>?
        var count = 0
        let status = rocr_detect_text_file(handle, path, options, &results, &count)
        guard status == 0 else { throw RustOError.recognitionFailed(status) }
        return try convertResults(results, count: count)
    }

    private func structuredData(_ handle: OpaquePointer, data: Data, options: String) throws -> [TextResult] {
        guard !data.isEmpty else { throw RustOError.invalidPath }
        var results: UnsafeMutablePointer<CTextResult>?
        var count = 0
        let status = data.withUnsafeBytes { bytes in
            rocr_detect_text_data(handle, bytes.baseAddress, data.count, options, &results, &count)
        }
        guard status == 0 else { throw RustOError.recognitionFailed(status) }
        return try convertResults(results, count: count)
    }

    private func convertResults(_ results: UnsafeMutablePointer<CTextResult>?, count: Int) throws -> [TextResult] {
        guard let results else {
            if count == 0 { return [] }
            throw RustOError.recognitionFailed(-1)
        }
        defer { rocr_free_results(results, count) }
        return (0..<count).map { index in
            let result = results[index]
            let points = [
                Point2D(x: result.box_x1, y: result.box_y1),
                Point2D(x: result.box_x2, y: result.box_y2),
                Point2D(x: result.box_x3, y: result.box_y3),
                Point2D(x: result.box_x4, y: result.box_y4),
            ]
            return TextResult(
                text: String(cString: result.text), score: result.score, boxPoints: points,
                frame: Frame(width: result.frame_width, height: result.frame_height, top: result.frame_top, left: result.frame_left)
            )
        }
    }

    private func validate(_ options: OcrRunOptions) throws {
        let validThreshold = { (value: Float?) in value == nil || (value!.isFinite && value! >= 0) }
        guard validThreshold(options.lineYThreshold), validThreshold(options.wordXThreshold),
              options.textScore == nil || (options.textScore!.isFinite && (0...1).contains(options.textScore!)) else {
            throw RustOError.serializationFailed
        }
        guard let preprocessing = options.preprocessing else { return }
        func positive(_ value: Float?) -> Bool { value == nil || (value!.isFinite && value! > 0) }
        guard positive(preprocessing.minHeight), positive(preprocessing.maxSideLen), positive(preprocessing.minSideLen),
              preprocessing.widthHeightRatio == nil || (preprocessing.widthHeightRatio!.isFinite && (preprocessing.widthHeightRatio! > 0 || preprocessing.widthHeightRatio! == -1)),
              preprocessing.minSideLen == nil || preprocessing.maxSideLen == nil || preprocessing.minSideLen! <= preprocessing.maxSideLen! else { throw RustOError.serializationFailed }
        guard let detection = preprocessing.detection else { return }
        guard detection.limitSideLen == nil || (detection.limitSideLen! > 0 && detection.limitSideLen! <= 32767),
              detection.limitType == nil || ["min", "max"].contains(detection.limitType!),
              detection.mean == nil || (detection.mean!.count == 3 && detection.mean!.allSatisfy { $0.isFinite }),
              detection.std == nil || (detection.std!.count == 3 && detection.std!.allSatisfy { $0.isFinite && $0 != 0 }) else { throw RustOError.serializationFailed }
        guard let postprocess = detection.postprocess else { return }
        guard postprocess.threshold == nil || (postprocess.threshold!.isFinite && (0...1).contains(postprocess.threshold!)),
              postprocess.boxThreshold == nil || (postprocess.boxThreshold!.isFinite && (0...1).contains(postprocess.boxThreshold!)),
              postprocess.maxCandidates == nil || postprocess.maxCandidates! >= 1,
              positive(postprocess.unclipRatio) else { throw RustOError.serializationFailed }
    }

    private static func resolveModelPath(_ filename: String) -> String? {
        if filename.hasPrefix("/") && FileManager.default.fileExists(atPath: filename) { return filename }
        let name = filename.replacingOccurrences(of: ".mnn", with: "").replacingOccurrences(of: ".txt", with: "")
        let ext = String(filename.split(separator: ".").last ?? "")
        let bundleNames = ["RustOModels", "RustoModels", "RustOModels_PPOCRv6_Tiny", "RustOModels_PPOCRv6_Small", "RustOModels_PPOCRv6_Medium", "RustOModels_PPOCRv5_Mobile", "RustOModels_PPOCRv5_Server", "RustOModels_PPOCRv4_Mobile", "RustOModels_PPOCRv4_Server"]
        for bundleName in bundleNames {
            if let path = Bundle.main.path(forResource: bundleName, ofType: "bundle"),
               let bundle = Bundle(path: path), let file = bundle.path(forResource: name, ofType: ext) { return file }
        }
        if let file = Bundle.main.path(forResource: filename, ofType: nil) { return file }
        let documents = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let file = (documents as NSString).appendingPathComponent(filename)
        return FileManager.default.fileExists(atPath: file) ? file : nil
    }

    private func resolveFilePath(_ value: String) -> String {
        var path = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if path.hasPrefix("file://"), let url = URL(string: path) { path = url.path }
        else if path.hasPrefix("file:") { path = String(path.dropFirst(5)) }
        return path.removingPercentEncoding ?? path
    }

    deinit { if let handle { rocr_free(handle) } }
}

// C API bridge
struct CTextResult {
    let text: UnsafeMutablePointer<CChar>
    let score: Float
    let box_x1, box_y1: Float
    let box_x2, box_y2: Float
    let box_x3, box_y3: Float
    let box_x4, box_y4: Float
    let frame_width, frame_height: Float
    let frame_top, frame_left: Float
}

@_silgen_name("rocr_initialize")
func rocr_initialize(_ configJson: String) -> OpaquePointer?

@_silgen_name("rocr_detect_text_file")
func rocr_detect_text_file(_ handle: OpaquePointer, _ imagePath: String, _ options: String, _ results: UnsafeMutablePointer<UnsafeMutablePointer<CTextResult>?>, _ count: UnsafeMutablePointer<Int>) -> Int32

@_silgen_name("rocr_detect_text_data")
func rocr_detect_text_data(_ handle: OpaquePointer, _ data: UnsafeRawPointer?, _ length: Int, _ options: String, _ results: UnsafeMutablePointer<UnsafeMutablePointer<CTextResult>?>, _ count: UnsafeMutablePointer<Int>) -> Int32

@_silgen_name("rocr_detect_text_file_spatial")
func rocr_detect_text_file_spatial(_ handle: OpaquePointer, _ imagePath: String, _ options: String) -> UnsafeMutablePointer<CChar>?

@_silgen_name("rocr_detect_text_data_spatial")
func rocr_detect_text_data_spatial(_ handle: OpaquePointer, _ data: UnsafeRawPointer?, _ length: Int, _ options: String) -> UnsafeMutablePointer<CChar>?

@_silgen_name("rocr_free_results")
func rocr_free_results(_ results: UnsafeMutablePointer<CTextResult>, _ count: Int)

@_silgen_name("rocr_free_string")
func rocr_free_string(_ str: UnsafeMutablePointer<CChar>)

@_silgen_name("rocr_free")
func rocr_free(_ handle: OpaquePointer)

@_silgen_name("rocr_version")
func rocr_version() -> UnsafePointer<CChar>?

File: packages/android/src/main/kotlin/com/byrizki/rusto/RustO.kt
package com.byrizki.rusto

import android.content.Context
import android.net.Uri
import org.json.JSONObject
import java.io.File
import java.net.URLDecoder

data class Point2D(val x: Float, val y: Float)

data class Frame(
    val width: Float,
    val height: Float,
    val top: Float,
    val left: Float
)

data class TextResult(
    val text: String,
    val score: Float,
    val boxPoints: List<Point2D>,
    val frame: Frame = run {
        val minX = boxPoints.minOfOrNull { it.x } ?: 0f
        val maxX = boxPoints.maxOfOrNull { it.x } ?: 0f
        val minY = boxPoints.minOfOrNull { it.y } ?: 0f
        val maxY = boxPoints.maxOfOrNull { it.y } ?: 0f
        Frame(width = maxX - minX, height = maxY - minY, top = minY, left = minX)
    }
)

data class DetectionConfig(
    val enabled: Boolean? = null,
    val modelPath: String? = null,
    val thresh: Float? = null,
    val boxThresh: Float? = null,
    val unclipRatio: Float? = null,
    val limitSideLen: Int? = null,
    val limitType: String? = null,
    val useDilation: Boolean? = null
)

data class RecognitionConfig(
    val enabled: Boolean? = null,
    val modelPath: String? = null,
    val dictPath: String? = null,
    val scoreThresh: Float? = null,
    val returnWordBox: Boolean? = null,
    val returnSingleCharBox: Boolean? = null
)

/**
 * Line Classification Configuration (CLS)
 * NOTE: Text line orientation classifier (180° rotation) is ONLY available on PP-OCRv4 and PP-OCRv5.
 */
data class ClassificationConfig(
    val enabled: Boolean? = null,
    val modelPath: String? = null,
    val thresh: Float? = null
)

data class OrientationConfig(
    val enabled: Boolean? = null,
    val modelPath: String? = null,
    val thresh: Float? = null
)

data class UnwarpConfig(
    val enabled: Boolean? = null,
    val modelPath: String? = null
)

data class LayoutConfig(
    val yThresholdMultiplier: Float? = null,
    val xThresholdMultiplier: Float? = null
)

data class InitializeConfig(
    val template: String? = null,
    val detection: DetectionConfig? = null,
    val recognition: RecognitionConfig? = null,
    val classification: ClassificationConfig? = null,
    val orientation: OrientationConfig? = null,
    val unwarp: UnwarpConfig? = null,
    val layout: LayoutConfig? = null
) {
    fun toJson(context: Context): String {
        val json = JSONObject()
        template?.let { json.put("template", it) }

        val detObj = JSONObject()
        val detPath = detection?.modelPath ?: "det.mnn"
        detObj.put("modelPath", resolveModel(context, detPath))
        detection?.enabled?.let { detObj.put("enabled", it) }
        detection?.thresh?.let { detObj.put("thresh", it.toDouble()) }
        detection?.boxThresh?.let { detObj.put("boxThresh", it.toDouble()) }
        detection?.unclipRatio?.let { detObj.put("unclipRatio", it.toDouble()) }
        detection?.limitSideLen?.let { detObj.put("limitSideLen", it) }
        detection?.limitType?.let { detObj.put("limitType", it) }
        detection?.useDilation?.let { detObj.put("useDilation", it) }
        json.put("detection", detObj)

        val recObj = JSONObject()
        val recPath = recognition?.modelPath ?: "rec.mnn"
        val dictPath = recognition?.dictPath ?: "dict.txt"
        recObj.put("modelPath", resolveModel(context, recPath))
        recObj.put("dictPath", resolveModel(context, dictPath))
        recognition?.enabled?.let { recObj.put("enabled", it) }
        recognition?.scoreThresh?.let { recObj.put("scoreThresh", it.toDouble()) }
        recognition?.returnWordBox?.let { recObj.put("returnWordBox", it) }
        recognition?.returnSingleCharBox?.let { recObj.put("returnSingleCharBox", it) }
        json.put("recognition", recObj)

        classification?.let { cls ->
            val clsObj = JSONObject()
            cls.enabled?.let { clsObj.put("enabled", it) }
            cls.modelPath?.let { clsObj.put("modelPath", resolveModel(context, it)) }
            cls.thresh?.let { clsObj.put("thresh", it.toDouble()) }
            if (clsObj.length() > 0) json.put("classification", clsObj)
        }

        orientation?.let { orient ->
            val orientObj = JSONObject()
            orient.enabled?.let { orientObj.put("enabled", it) }
            orient.modelPath?.let { orientObj.put("modelPath", resolveModel(context, it)) }
            orient.thresh?.let { orientObj.put("thresh", it.toDouble()) }
            if (orientObj.length() > 0) json.put("orientation", orientObj)
        }

        unwarp?.let { unw ->
            val unwObj = JSONObject()
            unw.enabled?.let { unwObj.put("enabled", it) }
            unw.modelPath?.let { unwObj.put("modelPath", resolveModel(context, it)) }
            if (unwObj.length() > 0) json.put("unwarp", unwObj)
        }

        layout?.let { lay ->
            val layObj = JSONObject()
            lay.yThresholdMultiplier?.let { layObj.put("yThresholdMultiplier", it.toDouble()) }
            lay.xThresholdMultiplier?.let { layObj.put("xThresholdMultiplier", it.toDouble()) }
            if (layObj.length() > 0) json.put("layout", layObj)
        }

        return json.toString()
    }

    private fun resolveModel(context: Context, pathOrAsset: String): String {
        val file = File(pathOrAsset)
        if (file.exists()) return file.absolutePath
        return RustO.copyAssetToCache(context, pathOrAsset)
    }
}

class RustOException(message: String) : Exception(message)

enum class OutputGranularity(val wireValue: String) {
    LINES("lines"), WORDS("words"), SPATIAL("spatial");
}

data class PostprocessRunOptions(
    val threshold: Float? = null,
    val boxThreshold: Float? = null,
    val maxCandidates: Int? = null,
    val unclipRatio: Float? = null,
    val useDilation: Boolean? = null,
)

data class DetectionRunOptions(
    val limitSideLen: Int? = null,
    val limitType: String? = null,
    val mean: FloatArray? = null,
    val std: FloatArray? = null,
)

data class OcrRunOptions(
    val output: OutputGranularity = OutputGranularity.LINES,
    val lineYThreshold: Float? = null,
    val wordXThreshold: Float? = null,
    val textScore: Float? = null,
    val classification: Boolean? = null,
    val orientation: Boolean? = null,
    val minHeight: Float? = null,
    val maxSideLen: Float? = null,
    val minSideLen: Float? = null,
    val widthHeightRatio: Float? = null,
    val detection: DetectionRunOptions? = null,
    val postprocess: PostprocessRunOptions? = null,
) {
    internal fun toJson(): String {
        require(lineYThreshold == null || lineYThreshold.isFinite() && lineYThreshold >= 0f) {
            "lineYThreshold must be finite and non-negative."
        }
        require(wordXThreshold == null || wordXThreshold.isFinite() && wordXThreshold >= 0f) {
            "wordXThreshold must be finite and non-negative."
        }
        require(textScore == null || textScore.isFinite() && textScore in 0f..1f) {
            "textScore must be finite and between 0 and 1."
        }
        validateRuntimeOverrides()
        return JSONObject().apply {
            put("output", output.wireValue)
            lineYThreshold?.let { put("lineYThreshold", it.toDouble()) }
            wordXThreshold?.let { put("wordXThreshold", it.toDouble()) }
            textScore?.let { put("textScore", it.toDouble()) }
            classification?.let { put("classification", it) }
            orientation?.let { put("orientation", it) }
            minHeight?.let { put("minHeight", it.toDouble()) }
            maxSideLen?.let { put("maxSideLen", it.toDouble()) }
            minSideLen?.let { put("minSideLen", it.toDouble()) }
            widthHeightRatio?.let { put("widthHeightRatio", it.toDouble()) }
            detection?.let { detection ->
                put("detection", JSONObject().apply {
                    detection.limitSideLen?.let { put("limitSideLen", it) }
                    detection.limitType?.let { put("limitType", it) }
                    detection.mean?.let { put("mean", it.toList()) }
                    detection.std?.let { put("std", it.toList()) }
                })
            }
            postprocess?.let { postprocess ->
                put("postprocess", JSONObject().apply {
                    postprocess.threshold?.let { put("threshold", it.toDouble()) }
                    postprocess.boxThreshold?.let { put("boxThreshold", it.toDouble()) }
                    postprocess.maxCandidates?.let { put("maxCandidates", it) }
                    postprocess.unclipRatio?.let { put("unclipRatio", it.toDouble()) }
                    postprocess.useDilation?.let { put("useDilation", it) }
                })
            }
        }.toString()
    }

    private fun validateRuntimeOverrides() {
        fun positive(name: String, number: Float?) = require(number == null || number.isFinite() && number > 0f) { "$name must be finite and greater than zero." }
        positive("minHeight", minHeight); positive("maxSideLen", maxSideLen); positive("minSideLen", minSideLen)
        require(widthHeightRatio == null || widthHeightRatio.isFinite() && (widthHeightRatio > 0f || widthHeightRatio == -1f)) { "widthHeightRatio must be positive or -1." }
        require(minSideLen == null || maxSideLen == null || minSideLen <= maxSideLen) { "minSideLen must not exceed maxSideLen." }
        detection?.let { detection ->
            require(detection.limitSideLen == null || detection.limitSideLen in 1..32767) { "limitSideLen must be between 1 and 32767." }
            require(detection.limitType == null || detection.limitType == "min" || detection.limitType == "max") { "limitType must be min or max." }
            for ((name, values, rejectZero) in listOf(Triple("mean", detection.mean, false), Triple("std", detection.std, true))) {
                require(values == null || values.size == 3 && values.all { it.isFinite() && (!rejectZero || it != 0f) }) { "$name must contain three valid numbers." }
            }
        }
        postprocess?.let { postprocess ->
            require(postprocess.threshold == null || postprocess.threshold.isFinite() && postprocess.threshold in 0f..1f) { "threshold must be between zero and one." }
            require(postprocess.boxThreshold == null || postprocess.boxThreshold.isFinite() && postprocess.boxThreshold in 0f..1f) { "boxThreshold must be between zero and one." }
            require(postprocess.maxCandidates == null || postprocess.maxCandidates >= 1) { "maxCandidates must be at least one." }
            require(postprocess.unclipRatio == null || postprocess.unclipRatio.isFinite() && postprocess.unclipRatio > 0f) { "unclipRatio must be finite and greater than zero." }
        }
    }
}

sealed interface ImageSource {
    data class Uri(val value: String) : ImageSource
    data class Bytes(val value: ByteArray) : ImageSource
}

sealed interface DetectTextResult {
    data class Structured(val items: List<TextResult>) : DetectTextResult
    data class Spatial(val text: String) : DetectTextResult
}

class RustO private constructor(
    private val context: Context,
    private var nativeHandle: Long,
) : AutoCloseable {

    companion object {
        init { System.loadLibrary("rusto") }

        @JvmStatic private external fun nativeInitialize(configJson: String): Long
        @JvmStatic private external fun nativeDetectTextFile(handle: Long, imagePath: String, optionsJson: String, resultsOut: LongArray, countOut: IntArray): Int
        @JvmStatic private external fun nativeDetectTextData(handle: Long, imageData: ByteArray, optionsJson: String, resultsOut: LongArray, countOut: IntArray): Int
        @JvmStatic private external fun nativeDetectTextFileSpatial(handle: Long, imagePath: String, optionsJson: String): String?
        @JvmStatic private external fun nativeDetectTextDataSpatial(handle: Long, imageData: ByteArray, optionsJson: String): String?
        @JvmStatic private external fun nativeGetResult(resultsPtr: Long, index: Int, textOut: Array<String>, scoreOut: FloatArray, boxOut: FloatArray)
        @JvmStatic private external fun nativeFreeResults(resultsPtr: Long, count: Int)
        @JvmStatic private external fun nativeFree(handle: Long)

        fun initialize(context: Context, config: InitializeConfig = InitializeConfig()): RustO {
            val handle = nativeInitialize(config.toJson(context))
            if (handle == 0L) throw RustOException("Failed to initialize RustO with config")
            return RustO(context.applicationContext, handle)
        }

        internal fun copyAssetToCache(context: Context, assetName: String): String {
            val cacheFile = File(context.cacheDir, assetName)
            if (cacheFile.exists() && cacheFile.length() > 0) return cacheFile.absolutePath
            cacheFile.parentFile?.mkdirs()
            val tmpFile = File(cacheFile.parent, "${cacheFile.name}.tmp")
            try {
                context.assets.open(assetName).use { input -> tmpFile.outputStream().use(input::copyTo) }
                tmpFile.renameTo(cacheFile)
            } catch (e: java.io.IOException) {
                tmpFile.delete()
                throw RustOException("Model asset not found or unreadable: '$assetName'. Cause: ${e.message}")
            }
            return cacheFile.absolutePath
        }
    }

    fun detectText(source: ImageSource, options: OcrRunOptions = OcrRunOptions()): DetectTextResult {
        check(nativeHandle != 0L) { "RustO is closed." }
        val optionsJson = options.toJson()
        return when (options.output) {
            OutputGranularity.SPATIAL -> DetectTextResult.Spatial(detectSpatial(source, optionsJson))
            OutputGranularity.LINES, OutputGranularity.WORDS -> DetectTextResult.Structured(detectStructured(source, optionsJson))
        }
    }

    private fun detectStructured(source: ImageSource, optionsJson: String): List<TextResult> {
        val resultsOut = LongArray(1)
        val countOut = IntArray(1)
        val status = when (source) {
            is ImageSource.Bytes -> nativeDetectTextData(nativeHandle, source.value, optionsJson, resultsOut, countOut)
            is ImageSource.Uri -> withUriInput(source.value) { path, bytes ->
                if (bytes != null) nativeDetectTextData(nativeHandle, bytes, optionsJson, resultsOut, countOut)
                else nativeDetectTextFile(nativeHandle, path!!, optionsJson, resultsOut, countOut)
            }
        }
        if (status != 0) throw RustOException("OCR detection failed with status: $status")
        try { return parseResults(resultsOut[0], countOut[0]) }
        finally { nativeFreeResults(resultsOut[0], countOut[0]) }
    }

    private fun detectSpatial(source: ImageSource, optionsJson: String): String = when (source) {
        is ImageSource.Bytes -> nativeDetectTextDataSpatial(nativeHandle, source.value, optionsJson)
        is ImageSource.Uri -> withUriInput(source.value) { path, bytes ->
            if (bytes != null) nativeDetectTextDataSpatial(nativeHandle, bytes, optionsJson)
            else nativeDetectTextFileSpatial(nativeHandle, path!!, optionsJson)
        }
    } ?: throw RustOException("OCR detection failed")

    private inline fun <T> withUriInput(value: String, action: (String?, ByteArray?) -> T): T {
        require(value.isNotBlank()) { "Image URI must be non-empty." }
        if (value.startsWith("content://")) {
            val bytes = context.contentResolver.openInputStream(Uri.parse(value))?.use { it.readBytes() }
                ?: throw RustOException("Failed to open URI: $value")
            return action(null, bytes)
        }
        var path = value.trim()
        if (path.startsWith("file://")) path = Uri.parse(path).path ?: path.removePrefix("file://")
        else if (path.startsWith("file:")) path = path.removePrefix("file:")
        path = try { URLDecoder.decode(path, "UTF-8") } catch (e: Exception) { path }
        require(File(path).isFile) { "Image file not found: $value" }
        return action(path, null)
    }

    private fun parseResults(resultsPtr: Long, count: Int): List<TextResult> {
        val results = ArrayList<TextResult>(count)
        val textOut = arrayOf("")
        val scoreOut = FloatArray(1)
        val boxOut = FloatArray(8)
        for (i in 0 until count) {
            nativeGetResult(resultsPtr, i, textOut, scoreOut, boxOut)
            val points = listOf(Point2D(boxOut[0], boxOut[1]), Point2D(boxOut[2], boxOut[3]), Point2D(boxOut[4], boxOut[5]), Point2D(boxOut[6], boxOut[7]))
            results += TextResult(textOut[0], scoreOut[0], points)
        }
        return results
    }

    override fun close() {
        if (nativeHandle != 0L) {
            nativeFree(nativeHandle)
            nativeHandle = 0L
        }
    }
}
